File: apps/android/feature/insights/domain/src/main/java/com/scribblefit/feature/insights/domain/usecase/GetVolumeInsightsUseCase.kt
package com.scribblefit.feature.insights.domain.usecase

import com.scribblefit.core.model.CurrentDate
import com.scribblefit.feature.insights.domain.model.VolumeDataPoint
import com.scribblefit.feature.insights.domain.repository.InsightsRepository
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flowOn

class GetVolumeInsightsUseCase(
    private val repository: InsightsRepository,
    private val coroutineDispatcher: CoroutineDispatcher
) {
    operator fun invoke(
        startDate: CurrentDate,
        endDate: CurrentDate
    ): Flow<List<VolumeDataPoint>> =
        repository.getVolumeInsights(
            startDate = startDate.millis,
            endDate = endDate.millis
        ).flowOn(coroutineDispatcher)
}

File: apps/ios/ScribbleFit/Sources/Domain/UseCases/GetVolumeInsightsUseCase.swift
import Foundation

public struct GetVolumeInsightsUseCase: Sendable {
    private let repository: InsightsRepository

    public init(repository: InsightsRepository) {
        self.repository = repository
    }

    public func execute(startDate: Date, endDate: Date) -> AsyncStream<[VolumeDataPoint]> {
        repository.getVolumeInsights(startDate: startDate, endDate: endDate)
    }
}
